### ISSUE TITLE
Can we have a alternative authentication way for policy-controller to pull the image for verification ?

### ISSUE THREAD

**Issue by MageshSrinivasulu** [sigstore/policy-controller#375]
My understanding is policy-controller will be using **_imagePullSecrets_** in the deployment manifest to pull the image from the private repository to verify its signature before deployment.

Is it possible to instruct policy-controller to use a secret that's not mentioned in the deployment manifest to pull the image?

We can have some fields in the cluster image policy which say. When you are trying to pull an image from a particular repository then use this secret for pulling the images

In that way, we don't need to rely on the secret provided in the deployment manifest.

Further, it will help in scenarios like where we have configuration embedded with the infrastructure / Kubernetes cluster specific to the cloud provider where deployment won't need imagePullSecrets for pulling the image since authentication happens based on the configurations we have already made. 

Baseline : Can we make policy-controller to use the secret through clusterimagepolicy since it already has the wildcard to match the private registry we can instruct it to use the secrets appropriately

**Comment by MageshSrinivasulu** [comment]
@hectorj2f Can you comment on this ?

**Comment by MageshSrinivasulu** [comment]
@hector Thanks for pointing me to this https://github.com/sigstore/policy-controller#configure-signaturepullsecrets I have tested this but have a few queries or you can we consider it as a feature requirement 

Can **oci** be made a not mandatory field? Because currently, we need to explicitly mention the path to the signature. Current working is not polymorphic more like static

If we don't mention oci it should fall back to the image path that is defined in the deployment manifest. 

Through this, we can decouple the secret that is required for image signature validation. So that cosign doesn't need to depend on the secret mentioned in the deployment manifest. 

image : example.com/path1@sha256:123

oci: example.com/path1 ( This path should dynamically get auto-populate for each image during its signature validation if we don't mention oci explicitly. We can achieve the above-mentioned logic if it is made possible )

This approach will be beneficial for people who don't want to rely on the secret mentioned in the deployment manifest.

**Comment by hectorj2f** [comment]
> Can oci be made a not mandatory field? Because currently, we need to explicitly mention the path to the signature. Current working is not polymorphic more like static

It is not a required field. Could you elaborate more about this ?

> If we don't mention oci it should fall back to the image path that is defined in the deployment manifest.

That is how it works today, and what I mentioned about the k8schain https://github.com/google/go-containerregistry/blob/main/pkg/authn/kubernetes/keychain.go#L52.

> oci: example.com/path1 ( This path should dynamically get auto-populate for each image during its signature validation if we don't mention oci explicitly. We can achieve the above-mentioned logic if it is made possible )

That is how it works. The OCI was an alternative based on your initial question about explicitly specifying a secret to use instead of relying on the one of the SA of the pod/deployment/ etc...

**Comment by MageshSrinivasulu** [comment]
**CURRENT:** 

Yes OCI is an alternative but it's restrictive at the same time. We need to mention the signature path explicitly, generally signature will be on the same path as that of the image. In that case, explicitly mentioning one is making things static. If we don't mention OCI is should consider the path of the image as the signature path as well.

```
spec:
  authorities:
  - key:
      secretRef:
        name: cosigned-public-key
    name: authority-0
    source:
      - oci: test.azurecr.io # THIS FIELD IS MANDATORY, WE CAN DEPLOY CLUSTER IMAGE POLICY WITH SOURCE SECTION 
                                         WITHOUT OCI MENTIONED RIGHT?
        signaturePullSecrets:
        - name: acrdev 
  images:
  - glob: test.azurecr.io/**
```

**EXPECTED:**

Can we make changes so the OCI won't be a mandatory field? I am expecting something like this below. 

```
spec:
  authorities:
  - key:
      secretRef:
        name: cosigned-public-key
    name: authority-0
    source:
      -  signaturePullSecrets:  # IF WE DON'T MENTION OCI, IMAGE PATH SHOULD BE TAKE FROM DEPLOYMENT MANIFEST 
                                                DYNAMICALLY
          - name: acrdev
  images:
  - glob: test.azurecr.io/** 
```

So that when we don't mention secrets in the service account. I could fall back to the cluster image policy where no OCI is mentioned. During that time policy controller needs to consider the image path mentioned in the deployment manifest to be the OCI.

Through that, we can make the working more dynamic. Since we have already mentioned glob, we can also pattern-match the image. So entire container registry can be matched with a single secret. We don't need to rely on a service account as well.

All I am requesting is policy controller should have configuration capability for matching all the sub-paths in the container registry with a single secret. Policy controller should use the secrets that were specifically created for it rather than relying on the service account secrets or image pull secrets in the manifest

**Comment by MageshSrinivasulu** [comment]
@hectorj2f Any feedback on the above explanation?

**Comment by hectorj2f** [comment]
@vaikas @mattmoor Do you have any better idea to solve this issue ?

**Comment by vaikas** [comment]
This seems like a reasonable approach to me.
https://github.com/sigstore/policy-controller/issues/375#issuecomment-1313119936

**Comment by MageshSrinivasulu** [comment]
@hectorj2f Can we change the label of the thread to **feature**? Since everyone is agreeing on the point. Keep me posted looking forward to test it once the changes are ready

**Comment by hectorj2f** [comment]
Yes, go for it!

**Comment by hectorj2f** [comment]
Closing this as resolved by #399
